internal/util/password/plain.go

```diff
@@ -0,0 +1,162 @@
+type plainParams struct {
+	t       uint32 // number of passes a.k.a. time
+	p       uint8  // degree of parallelism a.k.a. threads
+	m       uint32 // memory size in KiB
+	tagLen  uint32 // output length a.k.a. key length
+	saltLen int    // salt length
+}
+
+// fixedPlainParams represent fixed parameters for PLAIN authentication using Argon2id.
+//
+// We use values recommended by https://www.rfc-editor.org/rfc/rfc9106.html#section-4-6.2.
+// Some other sources, such as https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
+// or https://www.ietf.org/archive/id/draft-ietf-kitten-password-storage-04.html#name-argon2,
+// recommend lower parameters.
+// We also could automatically tweak parameters based on the available CPU and memory,
+// but let's keep it simple for now.
+var fixedPlainParams = plainParams{
+	t:       3,
+	p:       4,
+	m:       64 * 1024,
+	tagLen:  32,
+	saltLen: 16,
+}
+
+// plainHashParams hashes the password using Argon2id with the given salt and parameters,
+// and returns the document that should be stored and the hash.
+func plainHashParams(password string, salt []byte, params *plainParams) (*types.Document, []byte, error) {
+	if len(salt) != int(params.saltLen) {
+		return nil, nil, lazyerrors.Errorf("unexpected salt length: %d", len(salt))
+	}
+
+	hash := argon2.IDKey([]byte(password), salt, params.t, params.m, params.p, params.tagLen)
+
+	if len(hash) != int(params.tagLen) {
+		return nil, nil, lazyerrors.Errorf("unexpected hash length: %d", len(hash))
+	}
+
+	doc, err := types.NewDocument(
+		"algo", "argon2id",
+		"t", int32(params.t),
+		"p", int32(params.p),
+		"m", int32(params.m),
+		"hash", types.Binary{
+			B: hash,
+		},
+		"salt", types.Binary{
+			B: salt,
+		},
+	)
+	if err != nil {
+		return nil, nil, lazyerrors.Error(err)
+	}
+
+	return doc, hash, nil
+}
+
+// PlainHash hashes the password using Argon2id and returns the document that should be stored.
+func PlainHash(password string) (*types.Document, error) {
+	salt := make([]byte, fixedPlainParams.saltLen)
+	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	doc, _, err := plainHashParams(password, salt, &fixedPlainParams)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	return doc, nil
+}
+
+// plainVerifyParams verifies the password against the document returned by plainHashParams.
+//
+// Parameters are returned if they could be decoded from the document, even on error.
+func plainVerifyParams(password string, doc *types.Document) (*plainParams, error) {
+	v, _ := doc.Get("algo")
+	algo, _ := v.(string)
+
+	if algo != "argon2id" {
+		return nil, lazyerrors.Errorf("invalid algo: %q", algo)
+	}
+
+	v, _ = doc.Get("t")
+	t, _ := v.(int32)
+
+	v, _ = doc.Get("p")
+	p, _ := v.(int32)
+
+	v, _ = doc.Get("m")
+	m, _ := v.(int32)
+
+	v, _ = doc.Get("hash")
+	hash, _ := v.(types.Binary)
+
+	v, _ = doc.Get("salt")
+	salt, _ := v.(types.Binary)
```

Comment:
Is it okay if it panics due to incompatible type? Asking because we might want to be more strict here.

Reply:
All those values can't be zero/empty, so the normal error will be returned below.

Reply:
oh, now I've noticed that _ was used.